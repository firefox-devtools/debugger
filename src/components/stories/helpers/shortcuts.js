import PropTypes from "prop-types";
import { Component } from "react";

import { KeyShortcuts } from "devtools-modules";
const shortcuts = new KeyShortcuts({ window });

class Shortcuts extends Component {
  getChildContext() {
    return { shortcuts };
  }

  render() {
    return this.props.children;
  }
}

Shortcuts.childContextTypes = { shortcuts: PropTypes.object };
Shortcuts.propTypes = {
  children: PropTypes.object.isRequired
};

export default Shortcuts;
