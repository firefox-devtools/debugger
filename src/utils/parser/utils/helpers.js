// @flow

import * as t from "babel-types";
import type { NodePath, Node } from "babel-traverse";

import type { AstLocation, AstPosition } from "../types";

export function isLexicalScope(path: NodePath) {
  return t.isBlockStatement(path) || isFunction(path) || t.isProgram(path);
}

export function isFunction(path: NodePath) {
  return (
    t.isFunction(path) ||
    t.isArrowFunctionExpression(path) ||
    t.isObjectMethod(path) ||
    t.isClassMethod(path)
  );
}

export function isVariable(path: NodePath) {
  return (
    t.isVariableDeclaration(path) ||
    (isFunction(path) && path.node.params.length) ||
    (t.isObjectProperty(path) && !isFunction(path.node.value))
  );
}

export function getMemberExpression(root: Node) {
  function _getMemberExpression(node, expr) {
    if (t.isMemberExpression(node)) {
      expr = [node.property.name].concat(expr);
      return _getMemberExpression(node.object, expr);
    }

    if (t.isCallExpression(node)) {
      return [];
    }

    if (t.isThisExpression(node)) {
      return ["this"].concat(expr);
    }

    return [node.name].concat(expr);
  }

  const expr = _getMemberExpression(root, []);
  return expr.join(".");
}

export function containsPosition(a: AstLocation, b: AstPosition) {
  const startsBefore =
    a.start.line < b.line ||
    (a.start.line === b.line && a.start.column <= b.column);
  const endsAfter =
    a.end.line > b.line || (a.end.line === b.line && a.end.column >= b.column);

  return startsBefore && endsAfter;
}

export function containsLocation(a: AstLocation, b: AstLocation) {
  return containsPosition(a, b.start) && containsPosition(a, b.end);
}

export function nodeContainsPosition(node: Node, position: AstPosition) {
  const b = position.column ? position : { ...position, column: 0 };
  return containsPosition(node.loc, b);
}

export function nodeContainsLine(node: Node, position: AstPosition) {
  const a = node.loc;
  const b = position.column ? position : { ...position, column: 0 };

  console.log(a.start.line)
  const startsAtOrBefore = a.start.line === b.line;

  return startsAtOrBefore;
}
