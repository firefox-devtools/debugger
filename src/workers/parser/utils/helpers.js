// @flow

import * as t from "babel-types";
import type { NodePath, Node } from "babel-traverse";

export function isLexicalScope(path: NodePath) {
  return t.isBlockStatement(path) || isFunction(path) || t.isProgram(path);
}

export function isFunction(path: NodePath) {
  return (
    t.isFunction(path) ||
    t.isArrowFunctionExpression(path) ||
    t.isObjectMethod(path) ||
    t.isClassMethod(path) ||
    path.type === "MethodDefinition" ||
    (t.isClassProperty(path.parent) && t.isArrowFunctionExpression(path))
  );
}

export function isAwaitExpression(path: NodePath) {
  return (
    t.isAwaitExpression(path) ||
    t.isAwaitExpression(path.container.init) ||
    t.isAwaitExpression(path.parentPath)
  );
}

export function isYieldExpression(path: NodePath) {
  return (
    t.isYieldExpression(path) ||
    t.isYieldExpression(path.container.init) ||
    t.isYieldExpression(path.parentPath)
  );
}

export function isVariable(path: NodePath) {
  return (
    t.isVariableDeclaration(path) ||
    (isFunction(path) && path.node.params != null && path.node.params.length) ||
    (t.isObjectProperty(path) && !isFunction(path.node.value))
  );
}

export function getMemberExpression(root: Node) {
  function _getMemberExpression(node, expr) {
    if (t.isMemberExpression(node)) {
      expr = [node.property.name].concat(expr);
      return _getMemberExpression(node.object, expr);
    }

    if (t.isCallExpression(node)) {
      return [];
    }

    if (t.isThisExpression(node)) {
      return ["this"].concat(expr);
    }

    return [node.name].concat(expr);
  }

  const expr = _getMemberExpression(root, []);
  return expr.join(".");
}
