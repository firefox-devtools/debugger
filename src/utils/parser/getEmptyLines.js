import { uniq } from "lodash";
import { getAst } from "./utils/ast";

const commentTokens = ["CommentBlock", "CommentLine"];

export default function getEmptyLines(sourceToJS) {
  if (!sourceToJS) {
    return null;
  }

  const ast = getAst(sourceToJS);
  if (!ast || !ast.comments) {
    return [];
  }

  // The following sequence stores lines which have executable code
  // (contents other than comments or EOF, regardless of line position)
  const executableLines = ast.tokens
    .filter(token => {
      return (
        !commentTokens.includes(token.type) &&
        (!token.type || (token.type.label && token.type.label != "eof"))
      );
    })
    .map(token => token.loc.start.line - 1);
  executableLines = uniq(executableLines);

  // Populates a pre-filled array of every line number,
  // then removes lines which were found to be executable
  const fillRange = (start, end) => {
    return Array(end - start + 1).fill().map((item, index) => start + index);
  };
  const allLines = fillRange(0, ast.tokens[ast.tokens.length - 1].loc.end.line);

  return allLines.filter(i => !executableLines.includes(i));
}
