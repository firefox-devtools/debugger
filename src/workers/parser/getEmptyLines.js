import uniq from "lodash/uniq";
import difference from "lodash/difference";
import { getAst } from "./utils/ast";

const commentTokens = ["CommentBlock", "CommentLine"];

function fillRange(start, end) {
  return Array(end - start + 1)
    .fill()
    .map((item, index) => start + index);
}

// Populates a pre-filled array of every line number,
// then removes lines which were found to be executable
function getLines(ast) {
  return fillRange(1, ast.tokens[ast.tokens.length - 1].loc.end.line);
}

// The following sequence stores lines which have executable code
// (contents other than comments or EOF, regardless of line position)
function getExecutableLines(ast) {
  const lines = ast.tokens
    .filter(
      token =>
        !commentTokens.includes(token.type) &&
        (!token.type || (token.type.label && token.type.label != "eof"))
    )
    .map(token => token.loc.start.line);

  return uniq(lines);
}

export default function getEmptyLines(sourceToJS) {
  if (!sourceToJS) {
    return null;
  }

  const ast = getAst(sourceToJS);
  if (!ast || !ast.comments) {
    return [];
  }

  const executableLines = getExecutableLines(ast);
  const lines = getLines(ast);
  return difference(lines, executableLines);
}
