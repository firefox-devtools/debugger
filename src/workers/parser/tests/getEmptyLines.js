import { getSource } from "./helpers";
import getEmptyLines from "../getEmptyLines";

describe("getEmptyLines", () => {
  it("allSymbols", () => {
    expect(getEmptyLines(getSource("allSymbols"))).toMatchSnapshot();
  });
});
