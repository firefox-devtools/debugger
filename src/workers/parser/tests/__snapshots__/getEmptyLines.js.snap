// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getEmptyLines allSymbols 1`] = `
Array [
  2,
  6,
  8,
  18,
  21,
  26,
  31,
  32,
]
`;
